File: code/strings/brute-search.java
// Brute-force substring search: try the pattern at EVERY start position; on a
// mismatch, slide by ONE and restart the comparison. Worst case Theta(nm) —
// text characters get re-examined after a partial match.
static int bruteSearch(String t, String p) {
    int n = t.length(), m = p.length();
    for (int i = 0; i <= n - m; i++) {     // each start position
        int j = 0;
        while (j < m && t.charAt(i + j) == p.charAt(j)) j++;  // left to right
        if (j == m) return i;              // all m matched: found at i
    }
    return -1;                             // no occurrence
}

File: code/strings/brute-search.rs
// Brute-force substring search: try the pattern at EVERY start position; on a
// mismatch, slide by ONE and restart the comparison. Worst case O(nm) —
// text characters get re-examined after a partial match.
fn brute_search(t: &str, p: &str) -> i32 {
    let (t, p) = (t.as_bytes(), p.as_bytes());
    let (n, m) = (t.len(), p.len());
    if m > n { return -1; }
    for i in 0..=(n - m) {                 // each start position
        let mut j = 0;
        while j < m && t[i + j] == p[j] { j += 1; }   // compare left to right
        if j == m { return i as i32; }     // all m matched: found at i
    }
    -1                                     // no occurrence
}
